File: 1942/nessy/spritesheet.lua
local Sprite = {}

function Sprite:draw(point, scale)
	point = point or nessy.point()
	scale = scale or nessy.point(1, 1)

	local x, y = point:unpack()
	local sw, sh = scale:unpack()

	love.graphics.draw(self.spritesheet.raw, self.raw, x, y, 0, sw, sh)
end

local Spritesheet = {
	cache = {}
}

function Spritesheet:sprite(bounds)
	bounds = bounds or self.bounds
	local x, y, width, height = bounds:unpack()
	local sw, sh = self.bounds.size:unpack()

	local sprite = new(Sprite, {
		spritesheet = self,
		raw = love.graphics.newQuad(x, y, width, height, sw, sh),
		bounds = bounds:copy()
	})

	return sprite
end

function spritesheet(path)
	local raw = love.graphics.newImage(path)
	local spritesheet = new(Spritesheet, {
		raw = raw,
		bounds = nessy.rectangle(0, 0, raw:getWidth(), raw:getHeight())
	})

	return spritesheet
end

return spritesheet

File: src/nessy/spritesheet.js
nessy.Spritesheet = function(host) {
	var Spritesheet = function(image) {
		if (!(image instanceof Image)) throw "Not Image"
	
		this.raw = image
		this.bounds = new host.Rectangle(image.width, image.height)
	}
	
	Spritesheet.prototype = {
		sprite: function(bounds) {
			return new host.Sprite(this, bounds)
		}
	}
	
	return Spritesheet
}

nessy.Sprite = function(host) {
	var Sprite = function(spritesheet, bounds) {
		this.spritesheet = spritesheet
		this.bounds = bounds || spritesheet.bounds
	}
	
	Sprite.prototype = {
		draw: function(point, scale) {
			point = point || host.Point.zero
			scale = scale || host.Point.one
	
			host.graphics.drawImage(this.spritesheet.raw, point.x, point.y)
		}
	}
	
	return Sprite
}